k8s: keep CiliumNode labels synced with Node object

jibi opened the pull request:
This commit ensures that any update to the labels of a k8s Node object
is reflected in the corresponding CiliumNode one.

jibi commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -103,3 +107,27 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient, asyncCo
+	ciliumClient := k8s.CiliumClient()
+
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		ciliumNode, err := ciliumClient.CiliumV2().CiliumNodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
```
I initially followed the suggestion of using the k8s watcher node store:

```go
	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
		k8sNode, err := k.GetK8sNode(context.TODO(), nodeName)
		if err != nil {
			return err
		}

		nodeInterface := k8s.ConvertToNode(k8sNode)
		typesNode := nodeInterface.(*k8sTypes.Node)
		k8sNodeParsed := k8s.ParseNode(typesNode, source.Unspec)
		ciliumNode := k8sNodeParsed.ToCiliumNode()

		ciliumNode.ObjectMeta.Labels = labels

		_, err = ciliumClient.CiliumV2().CiliumNodes().Update(context.TODO(), ciliumNode, metav1.UpdateOptions{})

		return err
	})
```

but the `Update()` method was failing with `ciliumnodes.cilium.io "k8s1" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update`

I'm not sure if there's an easy way around this since the k8s watcher node store is storing `Node` objects rather than `CiliumNode` objects (and so I don't have access to the correct `resourceRevision` for the `CiliumNode` object I'm trying to update)

aanm replied:
We also have access to the cilium store that containers CiliumNodes. The idea would be that we would retrieve the labels from k8s node and update them into the a deep copy object retrieved from the CiliumNode store and only then use that object to update it with `_, err = ciliumClient.CiliumV2().CiliumNodes().Update(context.TODO(), ciliumNode, metav1.UpdateOptions{})`

jibi replied:
Switched to the `cache.Store` object returned by `informer.NewInformer()`

aanm commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -103,3 +107,27 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient, asyncCo
+func updateCiliumNodeLabels(nodeName string, labels map[string]string) error {
+	ciliumClient := k8s.CiliumClient()
+
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
Unfortunately we can't use this `retry.RetryOnConflict` because this is a blocking function and it will block the watcher. We need to use Cilium controllers instead.

christarazi commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -24,7 +24,7 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient, asyncCo
 	for {
 		swgNodes := lock.NewStoppableWaitGroup()
-		_, ciliumNodeInformer := informer.NewInformer(
+		ciliumNodeStore, ciliumNodeInformer := informer.NewInformer(
```
Does this mean we are now storing all CiliumNodes in each agent?

cc @aanm

aanm replied:
Yes, that is already the case.

aanm commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -76,6 +76,9 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient, asyncCo
 			k8s.ConvertToCiliumNode,
 		)
+
+		k.ciliumNodeStore = ciliumNodeStore
```
@jibi there's a small caveat with this.

As you can see this function is inside a for loop. In case Cilium connects to the kvstore, we can stop watching for `CiliumNodes` from Kubernetes (line 96) because the information that we care about is also available through the kvstore. However, the CiliumNodes in the kvstore don't have the labels information like we do have them in Kubernetes. Thus, if we really need to watch for `CiliumNodes` regardless if Cilium is or is not connected to the kvstore, the logic to stop the watcher, and the for loop, needs to be removed as well.

jibi replied:
Ack, I've removed the code that deals with the kvstore following what we are already doing for the `Node` watcher

aanm commented on pkg/k8s/watchers/node.go:
```diff
@@ -118,9 +123,54 @@ func (k *K8sWatcher) updateK8sNodeV1(oldK8sNode, newK8sNode *v1.Node) error {
+				ciliumNode := ciliumNodeInterface.(*cilium_v2.CiliumNode).DeepCopy()
+				ciliumNode.Labels = newNodeLabels
+
+				_, err = k8s.CiliumClient().CiliumV2().CiliumNodes().Update(context.TODO(), ciliumNode, metav1.UpdateOptions{})
```
`context.TODO()` can be replaced with the `ctx` from line 143

aanm commented on pkg/k8s/watchers/node.go:
```diff
@@ -118,9 +123,54 @@ func (k *K8sWatcher) updateK8sNodeV1(oldK8sNode, newK8sNode *v1.Node) error {
+					},
+				}
+
+				ciliumNodeInterface, exists, err := k.ciliumNodeStore.Get(pName)
```
we need to guarantee that `k.ciliumNodeStore` is set before we execute this function. This can easily be achieved by guaranteeing that the variable name is set before we execute this function.

aanm commented on pkg/k8s/watchers/node.go:
```diff
@@ -118,9 +123,58 @@ func (k *K8sWatcher) updateK8sNodeV1(oldK8sNode, newK8sNode *v1.Node) error {
+					},
+				}
+
+				if k.ciliumNodeStore == nil {
```
This will never be true since the interface will never be nil. See example in https://play.golang.org/p/h7-7jMLBxcM.

jibi replied:
I'm not sure I understand the issue (and so what I should do) as this check will return true until `k.ciliumNodeStore` is set in `pkg/k8s/watchers/cilium_node.go`:

```diff
diff --git a/pkg/k8s/watchers/cilium_node.go b/pkg/k8s/watchers/cilium_node.go
index d10a945db8..3f02dfd1be 100644
--- a/pkg/k8s/watchers/cilium_node.go
+++ b/pkg/k8s/watchers/cilium_node.go
@@ -16,7 +16,7 @@ import (
 )

 func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient) {
-       ciliumNodeStore, ciliumNodeInformer := informer.NewInformer(
+       _, ciliumNodeInformer := informer.NewInformer(
                cache.NewListWatchFromClient(ciliumNPClient.CiliumV2().RESTClient(),
                        cilium_v2.CNPluralName, v1.NamespaceAll, fields.Everything()),
                &cilium_v2.CiliumNode{},
@@ -69,8 +69,6 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient) {
                k8s.ConvertToCiliumNode,
        )

-       k.ciliumNodeStore = ciliumNodeStore
-
        k.blockWaitGroupToSyncResources(wait.NeverStop, nil, ciliumNodeInformer.HasSynced, k8sAPIGroupCiliumNodeV2)
        go ciliumNodeInformer.Run(wait.NeverStop)
        k.k8sAPIGroups.AddAPI(k8sAPIGroupCiliumNodeV2)
diff --git a/pkg/k8s/watchers/node.go b/pkg/k8s/watchers/node.go
index 793b6604c2..f7d5bfabe3 100644
--- a/pkg/k8s/watchers/node.go
+++ b/pkg/k8s/watchers/node.go
@@ -148,6 +148,7 @@ func (k *K8sWatcher) updateCiliumNodeLabels(nodeName string, newNodeLabels map[s
                                }

                                if k.ciliumNodeStore == nil {
+                                       log.Info("ciliumNodeStore is nil")
                                        return
                                }
```

```
vagrant@k8s1:~$ journalctl -u cilium | grep ciliumNodeStore
Oct 25 14:55:24 k8s1 cilium-agent[29772]: level=info msg="ciliumNodeStore is nil" subsys=k8s-watcher
```

> This can easily be achieved by guaranteeing that the variable name is set before we execute this function.

Otherwise I can move the `CiliumNode` entry further up in [(k *K8sWatcher)resourceGroups() ](https://github.com/cilium/cilium/blob/19cd3f3434dc246653a9de2dcbcc4381fb87c60f/pkg/k8s/watchers/watcher.go#L294-L296) to ensure its watcher is initialized before the `Node` one, but it feels like we are introducing an implicit dependency on the order of initialization of these resources

aanm commented on pkg/k8s/watchers/node.go:
```diff
@@ -118,9 +123,58 @@ func (k *K8sWatcher) updateK8sNodeV1(oldK8sNode, newK8sNode *v1.Node) error {
+				}
+
+				if k.ciliumNodeStore == nil {
+					return
```
We should return an error for this case because we want the controller to keep re-trying until the `ciliumNodeStore` is set.

jibi committed:
k8s: keep CiliumNode labels synced with Node object

This commit ensures that any update to the labels of a k8s Node object
is reflected in the corresponding CiliumNode one.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
